PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/ProgressUtils.cs

```diff
@@ -0,0 +1,331 @@
+							if (Thread.CurrentThread.ManagedThreadId != 1)
+							{
+								IAsyncResult handle = uiSynchronizationForm.BeginInvoke(getUIThreadDetails);
+								uiSynchronizationForm.EndInvoke(handle);
+							}
+							else
+								getUIThreadDetails();
+						}
+					}
+				}
+				return uiSynchronizationContext;
+			}
+		}
+
+		/// <summary>
+		/// InvokeLaterOnUIThread reentry protection
+		/// </summary>
+		static bool runningInvokeLaterOnUIThreadAction;
+
+		/// <summary>
+		/// If Reentry occurs while processing a InvokeLaterOnUIThread action, then the second action gets moved
+		/// into this List. When the original action is finished then the saved actions are reposted to the
+		/// synchronization context, and this list is cleared.
+		/// </summary>
+		static List<ThreadStart> postponedInvokeLaterOnUIThreadActions = new List<ThreadStart>();
+
+		/// <summary>
+		/// Gets whether or not the current thread is the main UI thread
+		/// </summary>
+		public static bool OnMainUiThread
+		{
+			get
+			{
+				// True if no UI context
+				SynchronizationContext context = UISynchronizationContext;
+				if (context == null)
+					return true;
+
+				if (uiSynchronizationThread != null && uiSynchronizationThread.Equals(Thread.CurrentThread))
+					return true;
+				// It's the same thread as our UI thread. (Fixes some tests that seem to change the current SynchronizationContext).
+
+				// Note: Assumes that synchronization context of UI thread is only on UI thread
+				return context.Equals(SynchronizationContext.Current);
+			}
+		}
+
+		/// <summary>
+		/// Invokes the specified action on the UI thread.  Blocking.
+		/// <para>NOTE: For this method to work, there needs to be an active message pump on the
+		/// UI thread (i.e. the action will only be run the next time the messages are handled).
+		/// Hopefully, if there is a visible form, there will be an active message pump.</para>
+		/// </summary>
+		/// <param name="action">the action to run on UI thread</param>
+		public static void InvokeOnUIThread(ThreadStart action)
+		{
+			InvokeOnUIThread(action, false);
+		}
+
+		/// <summary>
+		/// Invokes the specified action on the UI thread.  Blocking.
+		/// Does not interrupt calls to ExecuteOnSameThread.  Active message pump required.
+		/// </summary>
+		/// <param name="action">the action to run on UI thread</param>
+		public static void InvokeExclusivelyOnUIThread(ThreadStart action)
+		{
+			InvokeOnUIThread(action, true);
+		}
+
+		/// <summary>
+		/// Invokes the specified action on the UI thread.  Blocking.
+		/// Active message pump required.
+		/// </summary>
+		/// <param name="action">the action to run on UI thread</param>
+		/// <param name="exclusively">avoid overlap with ExecuteOnSameThread</param>
+		private static void InvokeOnUIThread(ThreadStart action, bool exclusively)
+		{
+			if (action == null)
+				return;
+
+			// Run directly if no UI context or if already on UI thread
+			if (OnMainUiThread)
```

Comment:
As mentioned in the comment here, this property checks more than just whether we're on the UI thread.  The property name and property comment should reflect this.